modules/generator/processor/spanmetrics/spanmetrics.go

```diff
@@ -115,17 +143,93 @@ func (p *Processor) aggregateMetricsForSpan(svcName string, rs *v1.Resource, spa
 	if p.Cfg.IntrinsicDimensions.StatusMessage {
 		labelValues = append(labelValues, span.GetStatus().GetMessage())
 	}
+
+	// if job is not present, remove label
+	if jobName != "" {
+		labelValues = append(labelValues, jobName)
+	} else {
+		p.labels = removeLabel(dimJob, p.labels)
+	}
+	// if instance is not present, remove label
+	if instanceID != "" {
+		labelValues = append(labelValues, instanceID)
+	} else {
+		p.labels = removeLabel(dimInstance, p.labels)
+	}
+
+	if p.Cfg.EnableTargetInfo {
+		// if job is not present, remove label
+		if jobName != "" {
+			targetInfoLabelValues = append(targetInfoLabelValues, jobName)
+		} else {
+			p.targetInfoLabels = removeLabel(dimJob, p.targetInfoLabels)
+		}
+	}
+
+	if p.Cfg.EnableTargetInfo {
+		// if instance is not present, remove label
+		if instanceID != "" {
+			targetInfoLabelValues = append(targetInfoLabelValues, instanceID)
+		} else {
+			p.targetInfoLabels = removeLabel(dimInstance, p.targetInfoLabels)
+		}
+	}
+
+	//update labels to remove job and/or instance if they're blank
+	p.spanMetricsCallsTotal.UpdateLabels(p.labels)
+	p.spanMetricsDurationSeconds.UpdateLabels(p.labels)
+	p.spanMetricsSizeTotal.UpdateLabels(p.labels)
+	p.spanMetricsTargetInfo.UpdateLabels(p.targetInfoLabels)
```

Comment:
Not sure if this is thread-safe. Batches can be concurrently pushed to a processor. If a set of traces has `instance_id`, and other doesn't, how does that affect all the labels that are being mutated in the metrics and the processor itself (eg. line 174)?

Reply:
good catch! i made some changes and added tests for this scenario